keps/sig-auth/4872-harden-kubelet-cert-validation/README.md

```diff
@@ -0,0 +1,386 @@
+In addition, we will log the error including the node name, so cluster administrators can identify which nodes are affected and need to reissue their certificates.
+
+We purposefully don't add the node name to the metric to avoid a high cardinality.
+The purpose of the metric is to easily/cheaply tell administrators if they can flip the feature on or not. If the answer is no (counter is greater than 0), the rest of the necessary information to detect the offending nodes will come from logs.
+
+Given that running the validation to feed the metric still has a cost, we won't run it if the validation is explicitly disabled with `--disable-kubelet-cert-cn-validation`.
+
+We will remove the metric once the feature is GA.
+
+> TODO: let's discuss this in the review. We could consider adding the node name to the metric or even keeping the metric post GA if it's valuable.
+
+### TLS insecure
+
+Currently, if the Kube-API server is not configured with a `--kubelet-certificate-authority` the TLS client for kubelet server will skip the server certificate validation.
+Additionally, `logs` requests allow to configure `InsecureSkipTLSVerifyBackend` per request to skip the server certificate validation.
+
+To align with this behavior, we won't execute the CN validation if `--kubelet-certificate-authority` is not set or if `InsecureSkipTLSVerifyBackend` is set to true.
+
+### Test Plan
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+##### Unit tests
+
+Unit tests will be added along with any new code introduced.
+
+Existing test coverage for the packages we anticipate modifying:
+
+- `k8s.io/kubernetes/pkg/kubelet/client`: `2024-10-07` - `28.2`
+- `k8s.io/client-go/transport`: `2024-10-07` - `59.4`
+
+##### Integration tests
+
+Integration tests will be added to ensure the following:
+* An error is returned if `--disable-kubelet-cert-cn-validation` is set but `KubeletCertCNValidation` feature flag is not enabled.
+* Validation for custom certificates works if feature flag is not enabled.
+* Validation for custom certificates works if feature flag enabled and `--disable-kubelet-cert-cn-validation` is set to true.
+* Validation for custom certificates fails if feature flag enabled and `--disable-kubelet-cert-cn-validation` is set to false or not set.
+* Validation for kubernetes issued certificates works if feature flag enabled and `--disable-kubelet-cert-cn-validation` is set to false or not set.
+
+##### e2e tests
+
+End-to-end tests won't be needed as unit and integration tests will cover all the scenarios.
+
+### Graduation Criteria
+
+#### Alpha
+
+* Add feature flag for gating usage, off by default
+* Add flag to disable extra validation
```

Comment:
remember you need to add the flag only if the feature flag is enabled, otherwise if the feature does not progress removing the flag can not be possible without breaking the users that are already setting it